### PR TITLE
[ZEND-557] Split_network join_network setup_network method refactoring

### PR DESCRIPTION
**Pull request by titusen**
Common Split_network method implentation move to "BitcoinTestFramework" class

### CODE REVIEW

File: qa/rpc-tests/test_framework/test_framework.py
Comment (ptagl) [inline qa/rpc-tests/test_framework/test_framework.py]: I like the idea of this PR and the approach of reducing the complexity of Python tests by making the connection functions of the framework more generic.

If I understand correctly, `setup_network()` still takes `split` as input parameter but the function always acts as it is always set to `False`. This may be a problem for some tests (like `qa/rpc-tests/sc_cert_addressindex.py` maybe) in which the network is split, then nodes are stopped and restarted and we expect to keep the previous state (split).

So, the question here is: can we find a way to keep using `self.is_network_split` to dynamically set up the network? In case, we would probably need to extend the current function with:
- a new input parameter for `setup_network()` to specify the node at the edge of the network (that is not connected to the next one)
- a class property to keep track of this node so that we can set up the network in the same way again in case of restart

Another possibility could be to completely get rid of `split` as input parameter, but these are just few ideas, let's wait for feedback also from the other reviewers and discuss what could be done next.

File: qa/rpc-tests/test_framework/test_framework.py
Comment (drgora) [inline qa/rpc-tests/test_framework/test_framework.py:91]: Just a minor thing here, but could we reuse `join_network` here, instead of reimplementing it?

File: qa/rpc-tests/test_framework/test_framework.py
Comment (drgora) [inline qa/rpc-tests/test_framework/test_framework.py]: Do we really need to comment this line out?
It looks safer to me with the `assert`, which might warn the user he is doing something unneeded

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (a-petrini) [inline qa/rpc-tests/sc_cert_addressindex.py]: this function appears twice in this class, which one should we keep?

File: qa/rpc-tests/sc_quality_nodes.py
Comment (a-petrini) [inline qa/rpc-tests/sc_quality_nodes.py]: I haven't re-run this test, but I think that removing the `sync_all()` at the end of the `setup_network()` is not safe

File: qa/rpc-tests/sc_cert_memcleanup_split.py
Comment (a-petrini) [inline qa/rpc-tests/sc_cert_memcleanup_split.py:239]: That's more a memo for me, actually, but I would like to investigate what is the purpose of this sleep and, eventually, if there could be a more reliable time-independent method for doing whatever it does, as it's present in other tests

Comment (titusen) [reply qa/rpc-tests/sc_cert_memcleanup_split.py:239]: There was a sleep in the join_network method of that class and without it the test tend to fail.

File: qa/rpc-tests/sc_quality_nodes.py
Comment (titusen) [reply qa/rpc-tests/sc_quality_nodes.py]: I removed this function, cause it can be inherited.

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (titusen) [reply qa/rpc-tests/sc_cert_addressindex.py]: Removed

File: qa/rpc-tests/sc_quality_nodes.py
Comment (a-petrini) [reply qa/rpc-tests/sc_quality_nodes.py]: Sorry, maybe my comment was a bit confusing, as I was referring to the last instruction of the `setup_network()` defined here: there was a `self.sync_all()` that has been removed, I don't know if it is intentional

Comment (titusen) [reply qa/rpc-tests/sc_quality_nodes.py]: The setup_network is derived from the base class. Now, it's only setup_node overidden.

Comment (a-petrini) [reply qa/rpc-tests/sc_quality_nodes.py]: Oh, I see! I was confused by the renamed function, thanks!

File: qa/rpc-tests/test_framework/test_framework.py
Comment (JackPiri) [inline qa/rpc-tests/test_framework/test_framework.py:69]: Possibly an assertion on `id` value can be added (e.g. for handling cases of out-of-range).
Similar comment applies to `join_network` and `join_network_sync`.

File: qa/rpc-tests/blockdelay.py
Comment (JackPiri) [inline qa/rpc-tests/blockdelay.py]: It seems to me there is no need to override `setup_nodes` with this specific function (having the exact same implementation).

File: qa/rpc-tests/blockdelay_2.py
Comment (JackPiri) [inline qa/rpc-tests/blockdelay_2.py:29]: I propose to remove both `join_network` and `join_network_2` from this test; the base class implementation can be used setting properly the `id` parameter (please note `sync_blocks` and `sync_mempools` are currently useless, being called on single-element-set, like `self.nodes[X:X+1]`).
Please replace accordingly calls to `join_network_2` with `join_network` with parameter `id=1` (lines 432, 572).

File: qa/rpc-tests/headers_01.py
Comment (JackPiri) [inline qa/rpc-tests/headers_01.py]: `split_network` can be removed (since it is not used).

File: qa/rpc-tests/sc_cert_nonceasing.py
Comment (JackPiri) [inline qa/rpc-tests/sc_cert_nonceasing.py]: `join_network` can be removed and the call at line 316 modified inserting parameter `id=0` (if the `time.sleep` call is really mandatory, it can be inserted just after, at line 317).

File: qa/rpc-tests/sc_csw_actcertdata.py
Comment (JackPiri) [inline qa/rpc-tests/sc_csw_actcertdata.py]: `split_network` can be removed (since it is not used).

File: qa/rpc-tests/sc_csw_fundrawtransaction.py
Comment (JackPiri) [inline qa/rpc-tests/sc_csw_fundrawtransaction.py]: `split_network` can be removed (since it is not used).

File: qa/rpc-tests/sc_ft_and_mbtr_fees_update.py
Comment (JackPiri) [inline qa/rpc-tests/sc_ft_and_mbtr_fees_update.py]: `join_network` can be removed (base impl can be used, with parameter `id=1`).

File: qa/rpc-tests/sc_fwd_maturity.py
Comment (JackPiri) [inline qa/rpc-tests/sc_fwd_maturity.py]: `split_network` can be removed (since it is not used).

**Review by JackPiri — CHANGES_REQUESTED** [review]
Good PR, it is a big effort toward making the test suite more clean and reusable.
I left some comments, it seems to me a good chance for pushing further with the refactoring; I think the PR needs to address them before being merged.
Additionally to the PR-involved-files comments, please consider:

- `sc_csw_balance_exceeding.py`: remove `split_network`/`join_network `(not even used),
- possibly remove unused imports (due to reference removal) on top of Pyhton tests.

Thanks,

File: qa/rpc-tests/test_framework/test_framework.py
Comment (a-petrini) [inline qa/rpc-tests/test_framework/test_framework.py]: I believe that `is_network_split` has to be set to `False` before calling `sync_all()`, since the behaviour of the latter function depends on this flag. In particular, with the current version of the code, mempool and chain synchronization happens amongst each group. Instead, we would like that after the execution of this function all the nodes have consistent mempool and blockchain.

Fixing the order will break / freeze `sc_invalidate.py` and `sc_csw_nullifier.py`, while logic in `sc_split.py` is not preserved.

File: qa/rpc-tests/sc_split.py
Comment (a-petrini) [inline qa/rpc-tests/sc_split.py]: This is somehow linked to my other comment, but in this test the old `join_network()` was not enforcing a sync on the nodes. As a matter of fact, in following `Checking mempools...` loop we are differentiating between the mempool groups. Still, the old `time.sleep(2)` plays a major role in not crashing the test... (see comment on `test_framework.py`)

File: qa/rpc-tests/test_framework/test_framework.py
Comment (titusen) [reply qa/rpc-tests/test_framework/test_framework.py:69]: If any method is called with out of range index, the "Unexpected exception caught during testing: list index out of range" is thrown and handled in test_framework.py.

```Split nodes (1)----x   x---(2)
Unexpected exception caught during testing: list index out of range
  File "/home/titusen/workspace/zen/qa/rpc-tests/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/home/titusen/workspace/zen/qa/rpc-tests/headers_06.py", line 130, in run_test
    self.split_network(10)
  File "/home/titusen/workspace/zen/qa/rpc-tests/test_framework/test_framework.py", line 71, in split_network
    disconnect_nodes(self.nodes[id], id + 1)
Stopping nodes
Cleaning up
Failed
!!! FAIL: headers_06.py !!!```

File: qa/rpc-tests/blockdelay.py
Comment (titusen) [reply qa/rpc-tests/blockdelay.py]: The default is 4 nodes, Here are 3.

File: qa/rpc-tests/headers_01.py
Comment (titusen) [reply qa/rpc-tests/headers_01.py]: Done

File: qa/rpc-tests/sc_csw_actcertdata.py
Comment (titusen) [reply qa/rpc-tests/sc_csw_actcertdata.py]: Done

File: qa/rpc-tests/sc_csw_fundrawtransaction.py
Comment (titusen) [reply qa/rpc-tests/sc_csw_fundrawtransaction.py]: Done

File: qa/rpc-tests/sc_ft_and_mbtr_fees_update.py
Comment (titusen) [reply qa/rpc-tests/sc_ft_and_mbtr_fees_update.py]: Done

File: qa/rpc-tests/blockdelay.py
Comment (JackPiri) [reply qa/rpc-tests/blockdelay.py]: Mhm, I see 4 nodes in the overriding:
```
def setup_nodes(self):
        self.nodes = start_nodes(4, self.options.tmpdir)
```
so it seems to me it's the same implementation.

**Review by JackPiri — APPROVED** [review]
The PR is fine to me.
Thanks

File: qa/rpc-tests/sc_cert_ceasing_split.py
Comment (a-petrini) [inline qa/rpc-tests/sc_cert_ceasing_split.py]: Assert at line 217 sporadically fails after this sync (verified with `run_until_fails.py` script). Using the `join_network_sync` made the test fail, so I suggest to revert back to the old one, for the meanwhile

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (a-petrini) [reply qa/rpc-tests/sc_cert_addressindex.py]: I believe you removed the wrong one: now `join_network()` is the function that stops and restarts the node (as defined here), while `split_network()` is the one in the base class, which just reconnect the nodes. It would be nice to preserve the symmetry of the functions originally defined.

As a side note, in the future, **please let the reviewer mark the conversation as resolved**, as issues like these may be overlooked.

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (a-petrini) [inline qa/rpc-tests/sc_cert_addressindex.py:368]: This will come handy after fixing the previous comment: teh subsequent call to `join_network()` at around line 440 has no parameter, so it will try to connect nodes 1 and 2, while the split created here are between nodes 0 and 1.

File: qa/rpc-tests/sbh_rpc_cmds.py
Comment (a-petrini) [inline qa/rpc-tests/sbh_rpc_cmds.py]: `split_network()` is not called anywhere in the test, so it can be removed

File: qa/rpc-tests/sc_invalidate.py
Comment (a-petrini) [inline qa/rpc-tests/sc_invalidate.py]: (see comment on `test_framework.py`)

File: qa/rpc-tests/sc_csw_nullifier.py
Comment (a-petrini) [inline qa/rpc-tests/sc_csw_nullifier.py]: (see comment on `test_framework.py`)

File: qa/rpc-tests/sc_quality_nodes.py
Comment (a-petrini) [inline qa/rpc-tests/sc_quality_nodes.py:213]: The new `join_network()` does not have the `time.sleep(2)`. Do you mind add it here, as done for `sc_cert_ceasing_split.py` and  `sc_cert_memcleanup_split.py`?

File: qa/rpc-tests/sc_quality_nodes.py
Comment (a-petrini) [inline qa/rpc-tests/sc_quality_nodes.py:312]: (see previous comment)

File: qa/rpc-tests/sc_csw_memcleanup_split.py
Comment (a-petrini) [inline qa/rpc-tests/sc_csw_memcleanup_split.py:236]: The new join_network() does not have the time.sleep(2). Do you mind add it here, as done for sc_cert_ceasing_split.py and sc_cert_memcleanup_split.py?

**Review by a-petrini — CHANGES_REQUESTED** [review]
Sorry for the delay, I've found these issues that should at least be discussed.
There are a few things that I have left out from the `headers_xx.py` test, that we could address in a separate ticket, since they need a major refactoring.

File: qa/rpc-tests/sc_split.py
Comment (titusen) [reply qa/rpc-tests/sc_split.py]: done

File: qa/rpc-tests/test_framework/test_framework.py
Comment (titusen) [reply qa/rpc-tests/test_framework/test_framework.py]: I thought that syncing entire network after join makes sense. I agree that it was this way.

File: qa/rpc-tests/sc_csw_memcleanup_split.py
Comment (titusen) [reply qa/rpc-tests/sc_csw_memcleanup_split.py:236]: Done

File: qa/rpc-tests/sc_cert_ceasing_split.py
Comment (titusen) [reply qa/rpc-tests/sc_cert_ceasing_split.py]: done

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (titusen) [reply qa/rpc-tests/sc_cert_addressindex.py:368]: Done

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (titusen) [reply qa/rpc-tests/sc_cert_addressindex.py]: done

File: qa/rpc-tests/sbh_rpc_cmds.py
Comment (titusen) [reply qa/rpc-tests/sbh_rpc_cmds.py]: Done

File: qa/rpc-tests/test_framework/test_framework.py
Comment (drgora) [inline qa/rpc-tests/test_framework/test_framework.py]: I agree with @a-petrini. This code, in its current form, is not actually synchronizing the entire network, but only nodes `0-1` and `2-3`. With the following modification, you should synchronize the entire network instead:
```suggestion
        self.is_network_split = False
        self.sync_all()
```

File: qa/rpc-tests/sc_invalidate.py
Comment (drgora) [inline qa/rpc-tests/sc_invalidate.py:202]: Would it be possible to use the new `join_network_sync()` instead of `time.sleep()` here?
```suggestion
        self.join_network_sync()
```

File: qa/rpc-tests/sc_cert_nonceasing.py
Comment (drgora) [inline qa/rpc-tests/sc_cert_nonceasing.py:310]: Here we synchronize already, couldn't we drop the `sleep()`?

File: qa/rpc-tests/test_framework/test_framework.py
Comment (drgora) [inline qa/rpc-tests/test_framework/test_framework.py:76]: This branch behaves as expected only if the network is split in the expected pair of nodes. Now that `split_network()` is more flexible, this might break.
Ideally, we should assess that the expected condition is met, and looking at `disconnect_nodes` I believe we already have all the needed components to implement a `nodes_disconnected()` helper function.
I understand this requires more effort, so feel free to ignore this comment  and keep it as a reminder for a future PR.

File: qa/rpc-tests/getchaintips.py
Comment (drgora) [inline qa/rpc-tests/getchaintips.py]: I guess the `??????` is a leftover?

File: qa/rpc-tests/getchaintips.py
Comment (drgora) [inline qa/rpc-tests/getchaintips.py:85]: This `join_network()` already executes a `sync_all()`. Is it possible to drop this `time.sleep(2)`?

File: qa/rpc-tests/headers_04.py
Comment (drgora) [inline qa/rpc-tests/headers_04.py:109]: In this and all other cases where we have a `join_network_sync()` followed by `time.sleep(2)`, could we remove `time.sleep(2)`?

File: qa/rpc-tests/sc_quality_nodes.py
Comment (drgora) [reply qa/rpc-tests/sc_quality_nodes.py:213]: Is that actually needed in this case? I believe the `sync_all()` at the following line should be a more reliable way of achieving the same state?

**Review by drgora — CHANGES_REQUESTED** [review]
Thank you for the PR! I really like how the new implementation avoids stopping and restarting nodes in `split_network` and `join_network`, this looks much more efficient.

I left few comments. In particular, I'd suggest to make sure if/why we need the many `time.sleep()` calls that we have. I see that most of the times we had them already, but this might be a great opportunity to get rid of them as much as possible. In general, and whenever possible, I would prefer having deterministic states, as example after a `sync_*` call, instead of relying on non-deterministic after sleep states.
This would make test more robust, and make the code simpler and more readable.

File: qa/rpc-tests/test_framework/test_framework.py
Comment (titusen) [reply qa/rpc-tests/test_framework/test_framework.py:91]: Will do

File: qa/rpc-tests/getchaintips.py
Comment (titusen) [reply qa/rpc-tests/getchaintips.py:85]: Moved to ZEND-616

File: qa/rpc-tests/headers_04.py
Comment (titusen) [reply qa/rpc-tests/headers_04.py:109]: Moved to ZEND-616](https://horizenlabs.atlassian.net/browse/ZEND-616)

File: qa/rpc-tests/sc_invalidate.py
Comment (titusen) [reply qa/rpc-tests/sc_invalidate.py:202]: Moved to ZEND-616](https://horizenlabs.atlassian.net/browse/ZEND-616)

File: qa/rpc-tests/sc_quality_nodes.py
Comment (titusen) [reply qa/rpc-tests/sc_quality_nodes.py:213]: Moved to ZEND-616](https://horizenlabs.atlassian.net/browse/ZEND-616)

File: qa/rpc-tests/sc_quality_nodes.py
Comment (titusen) [reply qa/rpc-tests/sc_quality_nodes.py:312]: Moved to ZEND-616](https://horizenlabs.atlassian.net/browse/ZEND-616)

File: qa/rpc-tests/test_framework/test_framework.py
Comment (titusen) [reply qa/rpc-tests/test_framework/test_framework.py:91]: Done

File: qa/rpc-tests/test_framework/test_framework.py
Comment (titusen) [reply qa/rpc-tests/test_framework/test_framework.py]: Done

File: qa/rpc-tests/test_framework/test_framework.py
Comment (titusen) [reply qa/rpc-tests/test_framework/test_framework.py:76]: Moved to ZEND-616](https://horizenlabs.atlassian.net/browse/ZEND-616)

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (a-petrini) [reply qa/rpc-tests/sc_cert_addressindex.py:368]: Nice, thanks!

File: qa/rpc-tests/sc_cert_addressindex.py
Comment (a-petrini) [reply qa/rpc-tests/sc_cert_addressindex.py]: thanks

File: qa/rpc-tests/test_framework/test_framework.py
Comment (a-petrini) [reply qa/rpc-tests/test_framework/test_framework.py]: Nice, thanks!

File: qa/rpc-tests/sc_cert_ceasing_split.py
Comment (a-petrini) [reply qa/rpc-tests/sc_cert_ceasing_split.py]: ok

File: qa/rpc-tests/sc_quality_nodes.py
Comment (a-petrini) [reply qa/rpc-tests/sc_quality_nodes.py:213]: @drgora yes, I agree that relying on a timed even is not safe, but I like the idea to investigate in another ticket why some tests do not need an explicit sync after join; therefore, my suggestion is to momentarily keep things as they were before, an move on with that refactoring in the other ticket

File: qa/rpc-tests/sc_csw_memcleanup_split.py
Comment (a-petrini) [reply qa/rpc-tests/sc_csw_memcleanup_split.py:236]: ok

File: qa/rpc-tests/sc_quality_nodes.py
Comment (a-petrini) [reply qa/rpc-tests/sc_quality_nodes.py:312]: thanks

**Review by a-petrini — APPROVED** [review]

File: qa/rpc-tests/sbh_rpc_cmds.py
Comment (a-petrini) [reply qa/rpc-tests/sbh_rpc_cmds.py]: thanks

File: qa/rpc-tests/sc_cert_nonceasing.py
Comment (titusen) [reply qa/rpc-tests/sc_cert_nonceasing.py:310]: In this test the sleep was in join. Moved to https://horizenlabs.atlassian.net/browse/ZEND-616

File: qa/rpc-tests/test_framework/test_framework.py
Comment (drgora) [reply qa/rpc-tests/test_framework/test_framework.py:91]: Great, thanks!

File: qa/rpc-tests/sc_quality_nodes.py
Comment (drgora) [reply qa/rpc-tests/sc_quality_nodes.py:213]: Fine for me, thanks

**Review by drgora — APPROVED** [review]
Looks good to me, thanks

**Review by ptagl — APPROVED** [review]

**Pull request merged**
